libbpf-tools/syscount.bpf.c

```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Anton Protopopov
+//
+// Based on syscount(8) from BCC by Sasha Goldshtein
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "syscount.h"
+
+const volatile bool count_by_process = false;
+const volatile bool measure_latency = false;
+const volatile bool filter_failed = false;
+const volatile int filter_errno = false;
+const volatile pid_t filter_pid = 0;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32);
+	__type(value, u64);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32);
+	__type(value, struct data_t);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} data SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_LRU_HASH);
+	__type(key, u32);
+	__type(value, __u8[COMM_NAME_SIZE]);
+} comm SEC(".maps");
+
+static __always_inline
+void *bpf_map_lookup_or_try_init(void *map, void *key, const void *init)
+{
+	void *val;
+	int err;
+
+	val = bpf_map_lookup_elem(map, key);
+	if (val)
+		return val;
+
+	err = bpf_map_update_elem(map, key, init, 0);
+	if (err)
+		return (void *) 0;
+
+	return bpf_map_lookup_elem(map, key);
+}
+
+static __always_inline
+void save_proc_name(pid_t pid)
+{
+	char buf[COMM_NAME_SIZE];
+	int err;
+
+	/* We should save the process name every time because a process can
+	 * change its comm. For example, when bash executes a binary program,
+	 * then first syscalls after fork, but before exec, run with the "bash"
+	 * comm, so if we cache this value here, then all processes created by
+	 * bash will be named "bash" */
+	err = bpf_get_current_comm(buf, sizeof(buf));
```

Comment:
Yeah, thanks, I will read process name instead (which makes me think about adding a `-t` option to count/display per-thread statistics).

So, for rewriting, I need to do this, because a process/thread can change its name, say, using `prctl` or after `exec`. (Probably it could be done in sync, like, when we obtain a process name once and then watching for syscalls which can change it, but this sounds like a hard thing to implement on practice for a debug tool?)